### PR TITLE
Cumulus 2191 async writes

### PR DESCRIPTION
**Pull request by Jkovarik**
**Summary:** Summary of changes

Addresses [CUMULUS-XX: Develop amazing new feature](https://bugs.earthdata.nasa.gov/browse/CUMULUS-XXX)

## Changes

* Detailed list or prose of changes
* ...

## PR Checklist

- [ ] Update CHANGELOG
- [ ] Unit tests
- [ ] Adhoc testing
- [ ] Integration tests

### CODE REVIEW

File: bamboo/docker-compose.yml
Comment (Jkovarik) [inline bamboo/docker-compose.yml:44]: Updated this (but not to 11, I've added a Jira ticket for it) while trying to run down a connection issue locally.     As tests pass and there are a few bugs fixed, leaving it in without objection.

File: lambdas/db-migration/src/migrations/20201014105030_create_async_operations_table.ts
Comment (Jkovarik) [inline lambdas/db-migration/src/migrations/20201014105030_create_async_operations_table.ts:21]: Added in #1947

File: package.json
Comment (Jkovarik) [inline package.json:21]: This is open for discussion - adding the async operations package results in a lerna link chain that exceeds a default search built into this module (see:  https://github.com/tclindner/npm-package-json-lint/issues/342 / 9e3c42bdb768d12ca60a3030747afb685fdca453 ) for more

File: packages/api/ecs/async-operation/index.js
Comment (Jkovarik) [inline packages/api/ecs/async-operation/index.js:186]: Nit: tidy up knex/id and just pass env to the method

File: packages/api/ecs/async-operation/index.js
Comment (Jkovarik) [inline packages/api/ecs/async-operation/index.js:193]: Nit: abstract this.

File: tf-modules/archive/async_operation.tf
Comment (Jkovarik) [inline tf-modules/archive/async_operation.tf:26]: Because the two new packages aren't published, and Docker won't let you bring symlinked lerna deps in, this requires a manual build until the release.

File: packages/async-operations/src/index.ts
Comment (Jkovarik) [inline packages/async-operations/src/index.ts:2]: This is a bit more than sub-optimal, but to stub out startAsyncOperation, we need to continue to use CommonJS exports or refactor significantly.    See slack discussion.

File: package.json
Comment (Jkovarik) [reply package.json:21]: I discussed this with Menno, and based on that conversation feel comfortable leaving this with a TD ticket in Jira to address.    If there's objections, we can have that conversation here/elsewhere.

File: packages/async-operations/src/index.ts
Comment (Jkovarik) [reply packages/async-operations/src/index.ts:2]: This was discussed and agreed that properly fixing the API endpoint test/logic/etc exceeds the scope for this ticket (see CUMULUS-2260 ) -- if there's an obvious proxyquire/setup for the tests that can be done cleanly in the API endpoints without significant refactor, I'm all 👂

File: packages/api/ecs/async-operation/index.js
Comment (Jkovarik) [reply packages/api/ecs/async-operation/index.js]: oops.  ceea0f760

File: lambdas/db-migration/package.json
Comment (markdboyd) [inline lambdas/db-migration/package.json:22]: Where is this used?

File: package.json
Comment (markdboyd) [reply package.json:21]: It looks like this supports globbing, so could it change to this instead?

`npmPkgJsonLint packages/* tasks/* example lambdas/*`

I wouldn't **think** that would recurse into packages/api/ecs/async-operation

**Review by markdboyd — CHANGES_REQUESTED** [review]
Looks great overall. A few questions, but a lot of comments are nits and/or questions

File: example/spec/serial/AsyncOperationRunnerNonExistentLambdaSpec.js
Comment (Jkovarik) [reply example/spec/serial/AsyncOperationRunnerNonExistentLambdaSpec.js:57]: This test was *not* passing, and it was not doing so because it was relying on the model.start

File: lambdas/db-migration/package.json
Comment (Jkovarik) [reply lambdas/db-migration/package.json:22]: Hmm, a fine question.   I updated this in response to my lerna issue, but assumed it was supposed to be there.   Looking.

File: package.json
Comment (Jkovarik) [reply package.json:21]: Sadly it appears to, I'm 99% sure I tested that exact approach (among others), but I'll give it another whirl to be sure once I've sorted the cyclic issue.

File: packages/api/ecs/async-operation/index.js
Comment (Jkovarik) [reply packages/api/ecs/async-operation/index.js:161]: Right - insert *fails* if the record exists.     This is the original reason I was looking into the upsert, but doing that doesn't necessarily have all the info needed to create a legit async record. 

The danger here is we write code that somehow starts this, but *doesn't* write the database record and this just silently succeeds (upsert where <can't find it>).  

We could probably validate the entry exists in the database and *fail* here if it doesn't if that's of concern.

Comment (Jkovarik) [reply packages/api/ecs/async-operation/index.js:161]: > but doing that doesn't necessarily have all the info needed to create a legit async record.

Also this is an existing issue with Dynamo, in theory.

File: packages/api/ecs/async-operation/tests/test-index.js
Comment (Jkovarik) [reply packages/api/ecs/async-operation/tests/test-index.js:66]: I'll remove, this was a debug convenience.

File: packages/api/ecs/async-operation/tests/test-index.js
Comment (Jkovarik) [reply packages/api/ecs/async-operation/tests/test-index.js:76]: L59

File: packages/api/endpoints/elasticsearch.js
Comment (Jkovarik) [reply packages/api/endpoints/elasticsearch.js:208]: Hmm, that's fair.    To be clear, are you suggesting having the unit manipulate process.env directly?

File: packages/api/endpoints/granules.js
Comment (Jkovarik) [reply packages/api/endpoints/granules.js:280]: Excellent catch.   I need to look into why this wasn't causing an integration test failure locally as if you look closely at `elasticsearch.js` the model was actually *in* the params object. 

4bd99eb35

File: packages/api/tests/models/test-async-operation-model.js
Comment (Jkovarik) [reply packages/api/tests/models/test-async-operation-model.js:93]: Indeed - I'm curious what the tablename would be there :)

File: packages/async-operations/src/async_operations.ts
Comment (Jkovarik) [reply packages/async-operations/src/async_operations.ts:60]: Right you are, will fix.

File: packages/async-operations/src/async_operations.ts
Comment (Jkovarik) [reply packages/async-operations/src/async_operations.ts:127]: Correct, noting to fix.

File: packages/async-operations/tests/test-async_operations.js
Comment (Jkovarik) [reply packages/async-operations/tests/test-async_operations.js]: Well that's embarrassing!   5eb909f64

File: packages/async-operations/tests/test-async_operations.js
Comment (Jkovarik) [reply packages/async-operations/tests/test-async_operations.js]: 24cf7477e

File: packages/async-operations/tests/test-async_operations.js
Comment (Jkovarik) [reply packages/async-operations/tests/test-async_operations.js:237]: I thought we were explicitly trying to avoid proxyquire when this sort of pattern will work   Left to my own, I'll probably just remove the comment, but if you feel strongly a proxyquire'd stub is preferable I'm 'ok' with it.

File: packages/async-operations/tests/test-async_operations.js
Comment (Jkovarik) [reply packages/async-operations/tests/test-async_operations.js:269]: Fair enough....though...  counter-nit: this will explicitly make sure it doesn't write more than one record.

File: packages/db/package.json
Comment (Jkovarik) [reply packages/db/package.json:3]: No, it probably isn't.     I'm unclear what the int tests have to do with it though.

File: packages/db/src/AsyncOperations.ts
Comment (Jkovarik) [reply packages/db/src/AsyncOperations.ts:1]: Good catch - I'd added this before bringing in your changes and hadn't updated it.

File: tf-modules/archive/async_operation.tf
Comment (Jkovarik) [reply tf-modules/archive/async_operation.tf:26]: Yes - I think we should get the PR approved, pass int tests, manually rebuild and then re-pass the integration tests with the updated image. 

I'll leave that up to the team as to if it should be a seperate PR.

File: example/package.json
Comment (Jkovarik) [reply example/package.json:50]: Yeah, this seemed like a reasonable point to make sure they were added for  version integrity reasons.  I can remove them if there's a reason to do so.

File: packages/api/ecs/async-operation/tests/test-index.js
Comment (Jkovarik) [reply packages/api/ecs/async-operation/tests/test-index.js]: 963e05473

File: packages/api/endpoints/elasticsearch.js
Comment (Jkovarik) [reply packages/api/endpoints/elasticsearch.js]: I think fafa53412 probably is the right call.  I'll update the others.

Comment (Jkovarik) [reply packages/api/endpoints/elasticsearch.js]: 65b44688d

File: packages/api/tests/endpoints/granules/test-bulk-delete.js
Comment (Jkovarik) [reply packages/api/tests/endpoints/granules/test-bulk-delete.js]: No, this was a comment re: the commonJS import.    But good point on .resolves: 

0ba574f7d

File: packages/api/tests/endpoints/granules/test-bulk-granules.js
Comment (Jkovarik) [reply packages/api/tests/endpoints/granules/test-bulk-granules.js:54]: 55610ee11

Comment (Jkovarik) [reply packages/api/tests/endpoints/granules/test-bulk-granules.js:54]: 55610ee11

File: packages/api/tests/endpoints/test-reconciliation-reports.js
Comment (Jkovarik) [reply packages/api/tests/endpoints/test-reconciliation-reports.js]: Hmm.     Well in this case, it was explicitly omitting the knexConfig, but I'm not sure I recall *why* that mattered.  This may be a victim of refactoring, but I agree it doesn't appear to be needed given the current state. 

216133a18 seems to be fine.

File: packages/api/tests/models/test-async-operation-model.js
Comment (Jkovarik) [reply packages/api/tests/models/test-async-operation-model.js]: Removed plus a bit more: 147dec8a7

File: packages/async-operations/src/async_operations.ts
Comment (Jkovarik) [reply packages/async-operations/src/async_operations.ts]: 🤦  yes.

b0f390528

File: packages/async-operations/src/index.ts
Comment (Jkovarik) [reply packages/async-operations/src/index.ts:2]: Agreed, I'm deeply unhappy with doing that.

File: packages/async-operations/tests/test-async_operations.js
Comment (Jkovarik) [reply packages/async-operations/tests/test-async_operations.js]: 619fdeb78

File: packages/async-operations/tests/test-async_operations.js
Comment (Jkovarik) [reply packages/async-operations/tests/test-async_operations.js]: 21d3c4b8d

File: packages/db/package.json
Comment (Jkovarik) [reply packages/db/package.json:3]: In either case it should be largely irrelevant until the feature merge.

File: packages/db/src/AsyncOperations.ts
Comment (Jkovarik) [reply packages/db/src/AsyncOperations.ts]: eef64f222

File: packages/async-operations/src/async_operations.ts
Comment (Jkovarik) [reply packages/async-operations/src/async_operations.ts]: Good point - that was a bit mangled.  

ff1f0d893

Comment (Jkovarik) [reply packages/async-operations/src/async_operations.ts]: Or rather: 6b6b2025e

File: packages/async-operations/src/async_operations.ts
Comment (Jkovarik) [reply packages/async-operations/src/async_operations.ts]: abc22d7eb450e9571275fc8c1e2c54620786fc16

File: packages/async-operations/src/async_operations.ts
Comment (Jkovarik) [reply packages/async-operations/src/async_operations.ts]: abc22d7eb450e9571275fc8c1e2c54620786fc16

File: packages/tea-map-cache/package.json
Comment (Jkovarik) [reply packages/tea-map-cache/package.json]: b88ae3b25

File: package.json
Comment (Jkovarik) [reply package.json]: *Apparently* some of the local build changes have 'sorted' the problem.   We'll see if it's reproducible in CI:

e9c023239

File: packages/async-operations/tests/test-async_operations.js
Comment (markdboyd) [reply packages/async-operations/tests/test-async_operations.js]: Fair point. You could have not made the change but since you already did 👍

File: packages/api/ecs/async-operation/index.js
Comment (markdboyd) [reply packages/api/ecs/async-operation/index.js:161]: OK, then this is fine

File: packages/api/ecs/async-operation/tests/test-index.js
Comment (markdboyd) [inline packages/api/ecs/async-operation/tests/test-index.js:42]: You can remove this

File: packages/api/ecs/async-operation/tests/test-index.js
Comment (markdboyd) [reply packages/api/ecs/async-operation/tests/test-index.js:71]: 🤦

File: packages/async-operations/src/async_operations.ts
Comment (markdboyd) [reply packages/async-operations/src/async_operations.ts]: The JSdoc for the `AsyncOperation` argument should also be updated, since it is specified as an object when it is really a class?

File: packages/api/endpoints/replays.js
Comment (markdboyd) [inline packages/api/endpoints/replays.js:6]: Is this still used in this file?

File: packages/api/tests/endpoints/granules/test-bulk-granules.js
Comment (markdboyd) [reply packages/api/tests/endpoints/granules/test-bulk-granules.js:54]: We should probably also remove the "TODO" note

File: packages/async-operations/package.json
Comment (markdboyd) [inline packages/async-operations/package.json:32]: I don't think the logger is used in here?

File: packages/async-operations/src/async_operations.ts
Comment (markdboyd) [inline packages/async-operations/src/async_operations.ts:112]: This type should change to a class, I think?

File: packages/async-operations/tests/test-async_operations.js
Comment (markdboyd) [inline packages/async-operations/tests/test-async_operations.js:86]: Remove this?

File: tf-modules/archive/async_operation.tf
Comment (markdboyd) [inline tf-modules/archive/async_operation.tf:26]: Reiterating: can we publish to cumuluss/async-operation once integration tests are passing here?

File: packages/async-operations/tests/test-async_operations.js
Comment (markdboyd) [inline packages/async-operations/tests/test-async_operations.js:23]: Is this necessary or does it need to be truly random?

File: lambdas/db-migration/package.json
Comment (markdboyd) [reply lambdas/db-migration/package.json]: It doesn't seem to be used that I can see

Comment (markdboyd) [reply lambdas/db-migration/package.json]: Actually, yes it is. My bad.

File: packages/async-operations/package.json
Comment (Jkovarik) [reply packages/async-operations/package.json]: 4ef7beb36

File: packages/async-operations/tests/test-async_operations.js
Comment (Jkovarik) [reply packages/async-operations/tests/test-async_operations.js]: 4ef7beb36

File: tf-modules/archive/async_operation.tf
Comment (Jkovarik) [reply tf-modules/archive/async_operation.tf:26]: Hmm, I thought I answered that:

https://github.com/nasa/cumulus/pull/1982#discussion_r517687688

File: tf-modules/archive/async_operation.tf
Comment (markdboyd) [reply tf-modules/archive/async_operation.tf]: If integration tests pass without being too painful, I'd probably vote on doing it in this PR, but we can see what happens

File: packages/async-operations/tests/test-async_operations.js
Comment (Jkovarik) [reply packages/async-operations/tests/test-async_operations.js]: No - it could be named something better. 

How about: 4e61ff4f9

File: packages/async-operations/package.json
Comment (Jkovarik) [reply packages/async-operations/package.json]: 4ef7beb36

File: packages/api/endpoints/replays.js
Comment (Jkovarik) [reply packages/api/endpoints/replays.js]: No - and it would fail lint - e73608ce3

**Review by markdboyd — APPROVED** [review]

File: packages/api/bin/serve.js
Comment (botanical) [inline packages/api/bin/serve.js:303]: this was probably a typo, but I think it should be `collectionModel` ?

Comment (Jkovarik) [reply packages/api/bin/serve.js:303]: Well *that* should fail a test.     :thinking face:

**Pull request merged**
